streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java

```diff
@@ -27,673 +25,314 @@
-                new KeyValue<>(3, 1.11f)  //Partition 0 - This will be the final result.
-        );
-
-        IntegrationTestUtils.produceKeyValuesSynchronously(LEFT_TABLE, table1ForeignKeyChange, LEFT_PROD_CONF, MOCK_TIME);
-        final List<KeyValue<Integer, String>> resultTwo = IntegrationTestUtils.readKeyValues(OUTPUT, CONSUMER_CONFIG, 15 * 1000L, Integer.MAX_VALUE);
-
-        final List<KeyValue<Integer, String>> expectedTwo = new LinkedList<>();
-        expectedTwo.add(new KeyValue<>(3, "value1=1.11,value2=10"));
-
-        assertArrayEquals(resultTwo.toArray(), expectedTwo.toArray());
-
-        //Ensure the state stores have the correct values within:
-        final Set<KeyValue<Integer, String>> expMatResults = new HashSet<>();
-        expMatResults.addAll(expected);
-        expMatResults.addAll(expectedTwo);
-        validateQueryableStoresContainExpectedKeyValues(expMatResults, currentMethodName);
-    }
-
-    private void createAndStartStreamsApplication(final String queryableStoreName, final boolean leftJoin) {
-        streamsConfig.put(StreamsConfig.APPLICATION_ID_CONFIG, "ktable-ktable-joinOnForeignKey-" + queryableStoreName);
-        streams = prepareTopology(queryableStoreName, leftJoin);
-        streamsTwo = prepareTopology(queryableStoreName, leftJoin);
-        streamsThree = prepareTopology(queryableStoreName, leftJoin);
-        streams.start();
-        streamsTwo.start();
-        streamsThree.start();
-    }
-
-    // These are hardwired into the test logic for readability sake.
-    // Do not change unless you want to change all the test results as well.
-    private ValueJoiner<Float, Long, String> joiner = (value1, value2) -> "value1=" + value1 + ",value2=" + value2;
-    //Do not change. See above comment.
-    private Function<Float, String> tableOneKeyExtractor = value -> Integer.toString((int) value.floatValue());
-
-    private void validateQueryableStoresContainExpectedKeyValues(final Set<KeyValue<Integer, String>> expectedResult,
-                                                                 final String queryableStoreName) {
-        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreOne = streams.store(queryableStoreName,
-                QueryableStoreTypes.keyValueStore());
-
-        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreTwo = streamsTwo.store(queryableStoreName,
-                QueryableStoreTypes.keyValueStore());
-
-        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreThree = streamsThree.store(queryableStoreName,
-                QueryableStoreTypes.keyValueStore());
-
-        // store only keeps last set of values, not entire stream of value changes
-        final Map<Integer, String> expectedInStore = new HashMap<>();
-        for (final KeyValue<Integer, String> expected : expectedResult) {
-            expectedInStore.put(expected.key, expected.value);
+    public void joinShouldProduceNullsWhenValueHasNonMatchingForeignKey() {
```

Comment:
I think we should include this test in the two from above and remove this test method -- it's not really a special case that needs it's own test method IMHO.

Reply:
Hmm. Not sure I agree. This case actually has its own top-level section in the KIP, namely what happens when you migrate from one non-existent FK to another. It seems worth explicitly calling out the truth table for this case.

The other tests are basically like, "what do I get if I add/remove a left record vs. a right record"? Sure, we could test all the permutations in one long test, but it's kind of hard to comprehend what such a test is doing after the fact.

Reply:
> Sure, we could test all the permutations in one long test, but it's kind of hard to comprehend what such a test is doing after the fact.

Not sure if I can follow. And yes, we should test all permutations according the table in the KIP IMHO.

Reply:
Sorry, re-reading that, it's a confusing statement. I just meant that tests that run through super long sequences of mutations are a little hard to understand later on.

Specifically, because the expected output is a function both of the current operation and the prior state the test has built up. So, if you make a change to Streams and an assertion fails halfway through a 500 line test method, it could take you an hour to figure out why the test even failed.

I think it would make more sense to pile all the test cases into one long method in tests where we actually have to run an embedded cluster, since it takes so long to start up and shut down, but for TTD tests, we can afford to have more targeted test methods.